packages/cli/src/services/cache/cache.service.ts

```diff
@@ -0,0 +1,338 @@
+
+			const { redisStoreUsingClient } = await import('@/services/cache/redis.cache-manager');
+			const redisStore = redisStoreUsingClient(redisClient, { ttl });
+
+			const redisCache = await caching(redisStore);
+
+			this.cache = { ...redisCache, kind: 'redis' };
+
+			return;
+		}
+
+		const maxSize = config.getEnv('cache.memory.maxSize');
+
+		const sizeCalculation = (item: unknown) => {
+			const str = jsonStringify(item, { replaceCircularRefs: true });
+			return new TextEncoder().encode(str).length;
+		};
+
+		const memoryCache = await caching('memory', { ttl, maxSize, sizeCalculation });
+
+		this.cache = { ...memoryCache, kind: 'memory' };
+	}
+
+	async reset() {
+		await this.cache.store.reset();
+	}
+
+	emit(event: CacheEvent, ...args: unknown[]) {
+		return super.emit(event, ...args);
+	}
+
+	isRedis() {
+		return this.cache.kind === 'redis';
+	}
+
+	isMemory() {
+		return this.cache.kind === 'memory';
+	}
+
+	// ----------------------------------
+	//             storing
+	// ----------------------------------
+
+	async set(key: string, value: unknown, ttl?: number) {
+		if (!this.cache) await this.init();
+
+		if (!key || !value) return;
+
+		if (this.cache.kind === 'redis' && !this.cache.store.isCacheable(value)) {
+			throw new UncacheableValueError(key);
+		}
+
+		await this.cache.store.set(key, value, ttl);
+	}
+
+	async setMany(keysValues: Array<[key: string, value: unknown]>, ttl?: number) {
+		if (!this.cache) await this.init();
+
+		if (keysValues.length === 0) return;
+
+		const truthyKeysValues = keysValues.filter(
+			([key, value]) => key?.length > 0 && value !== undefined && value !== null,
+		);
+
+		if (this.cache.kind === 'redis') {
+			for (const [key, value] of truthyKeysValues) {
+				if (!this.cache.store.isCacheable(value)) {
+					throw new UncacheableValueError(key);
+				}
+			}
+		}
+
+		await this.cache.store.mset(truthyKeysValues, ttl);
+	}
+
+	/**
+	 * Set or append to a [Redis hash](https://redis.io/docs/data-types/hashes/)
+	 * stored under a key in the cache. If in-memory, the hash is a regular JS object.
+	 */
+	async setHash(key: string, hash: Hash) {
+		if (!this.cache) await this.init();
+
+		if (!key?.length) return;
+
+		for (const hashKey in hash) {
+			if (hash[hashKey] === undefined || hash[hashKey] === null) return;
+		}
+
+		if (this.cache.kind === 'redis') {
+			await this.cache.store.hset(key, hash);
+			return;
+		}
+
+		const hashObject: Hash = (await this.get(key)) ?? {};
+
+		Object.assign(hashObject, hash);
+
+		await this.set(key, hashObject);
+	}
+
+	// ----------------------------------
+	//            retrieving
+	// ----------------------------------
+
+	async get<T = unknown>(
+		key: string,
+		{
+			fallbackValue,
+			refreshFn,
+		}: { fallbackValue?: T; refreshFn?: (key: string) => Promise<T | undefined> } = {},
+	) {
+		if (!this.cache) await this.init();
+
+		if (key?.length === 0) return;
+
+		const value = await this.cache.store.get<T>(key);
```

Comment:
To make sure I understand: 

- If the caller needs a primitive value, they'll call `get` (which uses `GET`).
- If they need the hash itself, they'll call `getHash` (which uses `HGETALL`).
- If they need a hash value they'll call `getHashValue` (which uses `HGET`).

If the caller calls `get` when they actually meant `getHash`, they'll see an unexpected result or failed retrieval _during development_. Do you mean we should be protecting the caller from mixing this up during development?

Reply:
Given we'll throw the original error, it should be fine.

This got me thinking whether we can protect or prevent this somehow, maybe tying specific keys to specific types, to prevent issues like this from happening (like saving a hash and trying to read using a regular `get`).

Future improvements I'd say.